realm/realm-library/src/main/java/io/realm/HandlerController.java

```diff
@@ -637,9 +637,17 @@ void addToRealmResults(RealmResults<? extends RealmModel> realmResults) {
         syncRealmResults.add(realmResultsWeakReference);
     }
 
-    // add to the list of RealmObject to be notified after a commit
-    <E extends RealmModel> void addToRealmObjects(E realmobject) {
-        realmObjects.put(new WeakReference<RealmObjectProxy>((RealmObjectProxy) realmobject), null);
+    // Add to the list of RealmObject to be notified after a commit.
+    // This method will check if the object exists in the list. It won't add the same object multiple times
+    <E extends RealmObjectProxy> void addToRealmObjects(E realmObject) {
+        for (WeakReference<RealmObjectProxy> ref : realmObjects.keySet()) {
```

Comment:
this is `O(n)` which will traverse each time all the `RealmObject` just to avoid duplicate, I suggest to use an alternative data structure (ex: check our `IdentitySet`, that guarantees `O(1)` on retrieve)